gqlopencensus/tracer.go

```diff
@@ -0,0 +1,140 @@
+package gqlopencensus
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/99designs/gqlgen/graphql"
+	"go.opencensus.io/trace"
+)
+
+var _ graphql.Tracer = (*tracerImpl)(nil)
+
+// New returns Tracer for OpenCensus.
+// see https://go.opencensus.io/trace
+func New(opts ...Option) graphql.Tracer {
+	tracer := &tracerImpl{}
+	cfg := &config{tracer}
+
+	for _, opt := range opts {
+		opt.apply(cfg)
+	}
+
+	return tracer
+}
+
+type tracerImpl struct {
+	startOperationExecutions     []func(ctx context.Context) context.Context
+	startFieldExecutions         []func(ctx context.Context, field graphql.CollectedField) context.Context
+	startFieldResolverExecutions []func(ctx context.Context, rc *graphql.ResolverContext) context.Context
+	startFieldChildExecutions    []func(ctx context.Context) context.Context
+	endFieldExecutions           []func(ctx context.Context)
+	endOperationExecutions       []func(ctx context.Context)
+}
+
+func (t *tracerImpl) StartOperationExecution(ctx context.Context) context.Context {
+	ctx, span := trace.StartSpan(ctx, operationName(ctx))
+	if !span.IsRecordingEvents() {
+		return ctx
+	}
+	requestContext := graphql.GetRequestContext(ctx)
+	span.AddAttributes(
+		trace.StringAttribute("request.query", requestContext.RawQuery),
+	)
+	for key, val := range requestContext.Variables {
+		span.AddAttributes(
+			trace.StringAttribute(fmt.Sprintf("request.variables.%s", key), fmt.Sprintf("%+v", val)),
+		)
+	}
+	for _, f := range t.startOperationExecutions {
+		ctx = f(ctx)
+	}
+
+	return ctx
+}
+
+func (t *tracerImpl) StartFieldExecution(ctx context.Context, field graphql.CollectedField) context.Context {
+	ctx, span := trace.StartSpan(ctx, field.ObjectDefinition.Name+"/"+field.Name)
+	if !span.IsRecordingEvents() {
```

Comment:
how does opencencus handle a request that wasn't sampled, but ends up causing an error? does try to promote it to being sampled?

Reply:
If our code doesn't have `if !span.IsRecordingEvents()` section, It works fine as well too.
OpenCensus doesn't change API behavior by sampling or not.
https://godoc.org/go.opencensus.io/trace#Span.IsRecordingEvents